spicy/src/rt/tests/parser.cc

```diff
@@ -114,4 +125,101 @@ TEST_CASE("registerParser") {
+    hilti::rt::done();
+    hilti::rt::init();
+
+    auto data = hilti::rt::ValueReference<hilti::rt::Stream>();
+
+    auto filters = hilti::rt::StrongReference<filter::detail::Filters>();
+
+    auto _waitForInputOrEod = [&](hilti::rt::resumable::Handle*) {
+        return detail::waitForInputOrEod(data, data->view(), 3, filters);
+    };
+
+    auto waitForInputOrEod = [&]() { return hilti::rt::fiber::execute(_waitForInputOrEod); };
+
+    SUBCASE("wait for nothing") {
+        // We can always successfully get "no data".
+        CHECK(detail::waitForInputOrEod(data, data->view(), 0, filters));
+    }
+
+    SUBCASE("not enough data") {
+        // `waitForInputOrEod` yields if not enough data is available. We
+        // can only wait from a `Resumable`.
+        CHECK_FALSE(waitForInputOrEod());
+        CHECK_THROWS_WITH_AS(_waitForInputOrEod(nullptr), "'yield' in non-suspendable context",
+                             const hilti::rt::RuntimeError&);
+    }
+
+    SUBCASE("enough data") {
+        // With enough data available we can get a result.
+        data->append("\x01\x02"_b);
+        REQUIRE_EQ(data->size(), 2);
+        CHECK_FALSE(waitForInputOrEod()); // Still need one more byte.
+
+        data->append("\x03");
+        REQUIRE_EQ(data->size(), 3);
+        const auto res = waitForInputOrEod();
+        REQUIRE(res);
+        CHECK(res.get<bool>());
+    }
+
+    SUBCASE("with filters") {
+        SUBCASE("empty filter list") {
+            filters = Vector<filter::detail::OneFilter>();
+
+            // Append enough data so this call would succeed if no filter was present.
+            data->append("\x01\x02\x03"_b);
+            REQUIRE_EQ(data->size(), 3);
+
+            // No filter present so we can get the available data directly.
+            const auto res = waitForInputOrEod();
+            REQUIRE(res);
+            CHECK(res.get<bool>());
+        }
+
+        SUBCASE("multiple filters") {
+            filters = Vector<filter::detail::OneFilter>();
+
+            bool called1 = false;
+            bool called2 = false;
+
+            filters->push_back(filter::detail::OneFilter({.resumable = [&](hilti::rt::resumable::Handle*) {
+                std::cerr << "FOO!\n";
+                called1 = true;
+            }}));
+            filters->push_back(filter::detail::OneFilter({.resumable = [&](hilti::rt::resumable::Handle*) {
+                std::cerr << "FOO!\n";
+                called2 = true;
+            }}));
+
+            REQUIRE_FALSE(called1);
+
+            // We trigger waiting for input with not enough data available and
+            // resume later as `waitForInputOrEod` would short-circuit were
+            // enough data available initally.
+            auto res = waitForInputOrEod();
+            data->append("\x01\x02\x03"_b);
+            res.resume();
+
+            CHECK(res);
+            CHECK(called1);
+            CHECK(called2);
+        }
+    }
+
+    SUBCASE("eod") {
```

Comment:
suggest to move this to above the filters; I had originally added comment there already to add an "eod" test. :-)